Lwm2m RD client update

jheiskan81 opened the pull request:
### RD Client bootstrap trigger update

Accept now from bootstrap trigger from any state expect ongoing bootstrap process.
Free also possible on going RD client message. There was a chance
that update message response change state and bootstrap process not started.

### LwM2M RD client message allocate update

Updated message handler that it free possible old allocated message.
Added message free also to RD stop and idle state.

### RD Client update registration timeout
Renamed LWM2M_RD_CLIENT_EVENT_REG_UPDATE_FAILURE to LWM2M_RD_CLIENT_EVENT_REG_TIMEOUT.
Changed reported new event type for registration timeout to LWM2M_RD_CLIENT_EVENT_REG_TIMEOUT
from LWM2M_RD_CLIENT_EVENT_REGISTRATION_FAILURE.
LWM2M_RD_CLIENT_EVENT_REGISTRATION_FAILURE should be only reported case when server reject by
response registration.

Commit:
net: lwm2m: LwM2M RD client message allocate update

Updated message handler that it free possible old allocated
message. Added message free also to RD stop and idle state.

Signed-off-by: Juha Heiskanen <juha.heiskanen@nordicsemi.no>

Commit:
net: lwm2m: RD Client bootstrap trigger update

Accept now from bootstrap trigger from any state expect ongoing
bootstrap process.
Free also possible on going RD client message. There was a chance
that update message response change state and bootstrap proces
not started.

Signed-off-by: Juha Heiskanen <juha.heiskanen@nordicsemi.no>

rlubos commented on subsys/net/lib/lwm2m/lwm2m_rd_client.c:
```diff
@@ -259,7 +266,7 @@ static void sm_handle_timeout_state(struct lwm2m_message *msg,
 	{
 		if (client.engine_state == ENGINE_REGISTRATION_SENT) {
-			event = LWM2M_RD_CLIENT_EVENT_REGISTRATION_FAILURE;
+			event = LWM2M_RD_CLIENT_EVENT_NETWORK_ERROR;
```
> LWM2M_RD_CLIENT_EVENT_REGISTRATION_FAILURE should be only reported case when server reject by response registration.

Why? How it's different from `LWM2M_RD_CLIENT_EVENT_REG_UPDATE_FAILURE` when both timeout and server rejectoin are handled the same? If it's needed to differentiate between timeout/rejection, perhaps a new event should be introduced?

jheiskan81 replied:
I was thinking to use that. I was seeing that server reject and timeout use same event but. Server rejected enter idle state but timeout continue healing at engine side. I think it make sense to use LWM2M_RD_CLIENT_EVENT_REG_UPDATE_FAILURE

rlubos commented on subsys/net/lib/lwm2m/lwm2m_rd_client.c:
```diff
@@ -404,18 +411,20 @@ int engine_trigger_bootstrap(void)
 	}
-
 	LOG_INF("Server Initiated Bootstrap");
+	/* Free ongoin possible message */
```
```suggestion
	/* Free ongoing possible message */
```

rlubos commented on subsys/net/lib/lwm2m/lwm2m_rd_client.c:
```diff
@@ -404,18 +411,20 @@ int engine_trigger_bootstrap(void)
-
 	LOG_INF("Server Initiated Bootstrap");
+	/* Free ongoin possible message */
+	lwm2m_reset_message(lwm2m_get_ongoing_rd_msg(), true);
```
Why not to use `rd_client_message_free()` helper?

jheiskan81 commented:
@rlubos I fix your comment's.

rlubos commented on doc/connectivity/networking/api/lwm2m.rst:
```diff
@@ -483,7 +483,7 @@ The events are prefixed with ``LWM2M_RD_CLIENT_EVENT_``.
      - REG_UPDATE_FAILURE
      - Registration update failed.
-       Occurs if there is a timeout during registration update.
+       Occurs if there is a timeout during registration or registration update.
```
I think my intentions were misunderstood here (sorry for not being clear enough). My point was, that we should be consistent with the events for Regsiter/Update/Deregister operations:
* either handle both timeout and error reply as the same kind of event for all operations (REGISTRATION_FAILURE/UPDATE_FAILURE/DEREGISTER_FAILURE) - that's how it's done right now,
* or introduce a new event for each operation which would allow to differentiate between timeout and error response (REGISTRATION_TIMEOUT/UPDATE_TIMEOUT/DEREGISTER_TIMEOUT).

If, from application perspective, it's important to differentiate between timeout and error reply, I think we should do the latter. But I'm strongly in favour of introducing new, specific events, instead of hiding the timeout event under one of the existing ones (which IMO could be more confusing than the current situation).

jheiskan81 replied:
OK. I create a new Event. I was wondering that we not break old application which missing a new event type.

jheiskan81 commented:
@rlubos I added a new Event for Registration timeout.

rlubos commented on include/zephyr/net/lwm2m.h:
```diff
@@ -1252,6 +1252,7 @@ enum lwm2m_rd_client_event {
 	LWM2M_RD_CLIENT_EVENT_BOOTSTRAP_TRANSFER_COMPLETE,
 	LWM2M_RD_CLIENT_EVENT_REGISTRATION_FAILURE,
 	LWM2M_RD_CLIENT_EVENT_REGISTRATION_COMPLETE,
+	LWM2M_RD_CLIENT_EVENT_REG_TIMEOUT_FAILURE,
```
I'd rather be consistent with other regsitration-related events naming, i. e. `LWM2M_RD_CLIENT_EVENT_REGISTRATION_TIMEOUT`

rlubos commented on doc/connectivity/networking/api/lwm2m.rst:
```diff
@@ -516,6 +520,11 @@ The events are prefixed with ``LWM2M_RD_CLIENT_EVENT_``.
        If sending a message fails, it will be retried.
        If the retry counter reaches its limits, this event will be triggered.
      - No actions needed, client will do a re-registrate automatically.
+   * - 12
```
This value doesn't seem to correspond to the actual enum value.

jheiskan81 replied:
Ah I need then change.

SeppoTakalo commented:
This is a bit confusing.
Why is there a separate state for UPDATE timeout and REGISTER timeout.
Bot require no action from device side. Both means, that we have no connection to server, yet.

Why we cannot refactor `REG_UPDATE_FAILURE` to be just `REG_TIMEOUT` and send same event from both of those.

rlubos commented:
> Bot require no action from device side. Both means, that we have no connection to server, yet.

If, by design, it's a no-op for the application, do we need to notify an event at all?

> I'm fine for REG_TIMEOUT It change just 1 exist event types. @rlubos any thought?

If we decide to keep timeout event, I'm fine with it.

SeppoTakalo commented:
> If, by design, it's a no-op for the application, do we need to notify an event at all?

I think some applications might want to know if they are connected right now, or not. For example if you have some UI indication that you actively talk to Cloud.

Or if you want to ask for a A-GPS assistance, you might want to know whether it is going to work at all.

Commit:
net: lwm2m: RD Client update registration timeout

Renamed LWM2M_RD_CLIENT_EVENT_REG_UPDATE_FAILURE to
LWM2M_RD_CLIENT_EVENT_REG_TIMEOUT.

Changed reported event type for registration timeout to
LWM2M_RD_CLIENT_EVENT_REG_TIMEOUT from
LWM2M_RD_CLIENT_EVENT_REGISTRATION_FAILURE.
LWM2M_RD_CLIENT_EVENT_REGISTRATION_FAILURE should be only
reported case when server reject by response registration.

Signed-off-by: Juha Heiskanen <juha.heiskanen@nordicsemi.no>

jheiskan81 commented:
@rlubos & @SeppoTakalo Please review. I fixed event name.

rlubos commented on include/zephyr/net/lwm2m.h:
```diff
@@ -1261,6 +1261,11 @@ enum lwm2m_rd_client_event {
+/**
+ *	Define for old event name keeping backward compability.
+ */
+#define LWM2M_RD_CLIENT_EVENT_REG_UPDATE_FAILURE LWM2M_RD_CLIENT_EVENT_REG_TIMEOUT
```
Could use `__DEPRECATED_MACRO` here so that we can remove the old name eventually.

jheiskan81 replied:
That's a good idea.

rlubos reviewed (dismissed):
Jus one nit.
